<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="UTF-8">
	<title>promise</title>
</head>
<body>
	
	<script>

	/*
	Promise的内部实现是一个状态机。Promise有三种状态：pending，resolved，rejected。当Promise刚创建完成时，处于pending状态；当Promise中的函数参数执行了resolve后，Promise由pending状态变成resolved状态；如果在Promise的函数参数中执行的不是resolve方法，而是reject方法，那么Promise会由pending状态变成rejected状态。

p2、p3刚创建完成时，控制台输出的这两台Promise都处于pending状态，但为什么p1是resolved状态呢？ 这是因为p1 的函数参数中执行的是一段同步代码，Promise刚创建完成，resolve方法就已经被调用了，因而紧跟着的输出显示p1是resolved状态。我们通过两个setTimeout函数，延迟1s后再次输出p2、p3的状态，此时p2、p3已经执行完成，状态分别变成resolved和rejected。
	 */
		var log = console.log

		var p1 = new Promise(function(resolve, reject){

			resolve(1);
		})

		var p2 = new Promise(function(resolve, reject){
			setTimeout(function(){
				resolve(2);
			},500)
		})

		var p3 = new Promise(function(resolve,reject){
			setTimeout(function(){
				reject(3)
			},500)
		})


		log(p1);
		log(p2);
		log(p3);

		setTimeout(function(){
			log(p1);
		},500);
		setTimeout(function(){
			log(p2);
		},500);
		setTimeout(function(){
			log(p3);
		},500);

		p1.then(function(value){
			log(value);
		})

		p2.then(function(value){
			log(value);
		})

		p3.catch(function(err){
			log(err);
		})
	</script>
</body>
</html>